pkg/scheduler/objects/node_collection.go

```diff
@@ -0,0 +1,245 @@
+/*
+ Licensed to the Apache Software Foundation (ASF) under one
+ or more contributor license agreements.  See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership.  The ASF licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package objects
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/google/btree"
+	"go.uber.org/zap"
+
+	"github.com/apache/incubator-yunikorn-core/pkg/log"
+)
+
+type NodeCollection interface {
+	AddNode(node *Node) error
+	RemoveNode(nodeID string) *Node
+	GetNode(nodeID string) *Node
+	GetNodeCount() int
+	GetNodes() []*Node
+	GetSchedulableNodes(excludeReserved bool) []*Node
+	GetSchedulableNodeIterator() NodeIterator
+	SetNodeSortingPolicy(policy NodeSortingPolicy)
+	GetNodeSortingPolicy() NodeSortingPolicy
+}
+
+type nodeRef struct {
+	node      *Node   // node reference
+	nodeScore float64 // node score
+}
+
+func (nr nodeRef) Less(than btree.Item) bool {
+	other, ok := than.(nodeRef)
+	if !ok {
+		return false
+	}
+	if nr.nodeScore < other.nodeScore {
+		return true
+	}
+	if other.nodeScore < nr.nodeScore {
+		return false
+	}
+	return nr.node.NodeID < other.node.NodeID
+}
+
+type baseNodeCollection struct {
+	Partition string // partition used with this collection
+
+	// Private fields need protection
+	nsp         NodeSortingPolicy   // node sorting policy
+	nodes       map[string]*nodeRef // nodes assigned to this collection
+	activeNodes *btree.BTree        // nodes which are active
+
+	sync.RWMutex
+}
+
+func (nc *baseNodeCollection) scoreNode(node *Node) float64 {
+	if nc.nsp == nil {
+		return 0
+	}
+	return nc.nsp.ScoreNode(node)
+}
+
+// Add a node to the collection by nodeID.
+func (nc *baseNodeCollection) AddNode(node *Node) error {
+	nc.Lock()
+	defer nc.Unlock()
+
+	if node == nil {
+		return fmt.Errorf("node cannot be nil")
+	}
+	if nc.nodes[node.NodeID] != nil {
+		return fmt.Errorf("partition %s has an existing node %s, node name must be unique", nc.Partition, node.NodeID)
+	}
+	// Node can be added to the system to allow processing of the allocations
+	node.AddListener(nc)
+	nref := nodeRef{
+		node:      node,
+		nodeScore: nc.scoreNode(node),
+	}
+	nc.nodes[node.NodeID] = &nref
+	if node.IsSchedulable() && !node.IsReserved() {
+		nc.activeNodes.ReplaceOrInsert(nref)
+	}
```

Comment:
This PR is focused on maintaining existing behavior but improving performance. I'd prefer to avoid behavior changes in the code until they are needed. Could we defer that to a future ticket?

Reply:
Make sense, let's take care of that in different JIRAs.